File: 0725-split-linked-list-in-parts/0725-split-linked-list-in-parts.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 * int val;
 * ListNode next;
 * ListNode() {}
 * ListNode(int val) { this.val = val; }
 * ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode[] splitListToParts(ListNode head, int k) {
        ListNode[] ans = new ListNode[k];
        ListNode curr = head;
        int cnt = 0;
        while (curr != null) {
            cnt++;
            curr = curr.next;
        }
        int step = cnt / k, rem = cnt % k;
        curr = head;
        for (int i = 0; i < k; i++) {
            ans[i] = head;
            cnt = step;
            if (rem > 0) {
                rem--;
                cnt++;
            }
            while (cnt > 0) {
                curr = head;
                if (head != null)
                    head = head.next;
                cnt--;
            }
            if (curr != null) curr.next = null;
        }
        return ans;
    }
}

File: 0725-split-linked-list-in-parts/0725-split-linked-list-in-parts.kt
/**
 * Example:
 * var li = ListNode(5)
 * var v = li.`val`
 * Definition for singly-linked list.
 * class ListNode(var `val`: Int) {
 *     var next: ListNode? = null
 * }
 */
class Solution {
    fun splitListToParts(head: ListNode?, k: Int): Array<ListNode?> {
        val ans: Array<ListNode?> = Array(k) { null };
        var dh: ListNode? = head;
        var curr: ListNode? = head;
        var cnt: Int = 0;
        while (curr != null) {
            cnt++;
            curr = curr?.next;
        }
        var step: Int = cnt / k; var rem: Int = cnt % k;
        curr = head;
        for (i in 0 until k) {
            ans[i] = dh;
            cnt = step;
            if (rem > 0) {
                rem--;
                cnt++;
            }
            while (cnt > 0) {
                curr = dh;
                if (dh != null)
                    dh = dh?.next;
                cnt--;
            }
            if (curr != null) curr?.next = null;
        } 
        return ans;
    }
}